effectful/handlers/llm/providers.py

```diff
@@ -92,36 +94,79 @@ def _(values: Sequence) -> OpenAIMessageContent:
         return [{"type": "text", "text": str(values)}]
 
 
+@format_value.register(types.FunctionType)  # type: ignore
+def _(obj: types.FunctionType) -> OpenAIMessageContent:
+    try:
+        source = textwrap.dedent(inspect.getsource(obj)).strip()
+    except (OSError, TypeError):
+        doc = obj.__doc__ or "No docstring"
+        source = f"# <function {obj.__name__}>\n# {doc}"
+    return [{"type": "text", "text": source}]
+
+
+@format_value.register(type)  # type: ignore
+def _(obj: type) -> OpenAIMessageContent:
+    try:
+        source = textwrap.dedent(inspect.getsource(obj)).strip()
+    except (OSError, TypeError):
+        doc = obj.__doc__ or "No docstring"
+        source = f"# <class {obj.__name__}>\n# {doc}"
+    return [{"type": "text", "text": source}]
+
+
+@format_value.register(types.ModuleType)  # type: ignore
+def _(obj: types.ModuleType) -> OpenAIMessageContent:
+    # Return empty for modules (skip in lexical context)
+    return []
+
+
 @dataclasses.dataclass
 class Tool[**P, T]:
     parameter_model: type[pydantic.BaseModel]
-    operation: Operation[P, T]
+    callable: "Operation[P, T] | Template[P, T]"
     name: str
+    description: str
 
     def serialise_return_value(self, value) -> OpenAIMessageContent:
         """Serializes a value returned by the function into a json format suitable for the OpenAI API."""
-        sig = inspect.signature(self.operation)
+        if isinstance(self.callable, Template):
+            sig = self.callable.__signature__
+        else:
+            sig = inspect.signature(self.callable)
         ret_ty = sig.return_annotation
         ret_ty_origin = typing.get_origin(ret_ty) or ret_ty
 
         return format_value.dispatch(ret_ty_origin)(value)  # type: ignore
 
     @classmethod
-    def of_operation(cls, op: Operation[P, T], name: str):
-        sig = inspect.signature(op)
-        hints = get_type_hints(op)
-        fields = {
-            param_name: hints.get(param_name, str) for param_name in sig.parameters
-        }
+    def of(cls, obj: "Operation[P, T] | Template[P, T]", name: str):
+        """Create a Tool from an Operation or Template."""
+        if isinstance(obj, Template):
```

Comment:
You shouldn't need two separate code paths here, the logic for `Template` should be exactly the same as the other case.